File: CPP/program/Tampilan.cpp
#pragma once

#include <iostream>
#include <string>
#include <vector>

#include "Car.cpp"
#include "Motorcycle.cpp"

using namespace std;

class Tampilan {
private:
    vector<int> kolom;
public:
    

    Tampilan(int jumlah){
        for(int i = 0;i < jumlah;i++){
            kolom.push_back(0);
            
        }
    }





    void cek_kolom_mobil(Car car){
        if(kolom[0] < car.getPlatNomor().length() + 2){
            kolom[0] = car.getPlatNomor().length() + 2;
        }
        if(kolom[1] < car.getMerk().length() + 2){
            kolom[1] = car.getMerk().length() + 2;
        }
        if(kolom[2] < car.getWarna().length() + 2){
            kolom[2] = car.getWarna().length() + 2;
        }
        if(kolom[3] < to_string(car.getTahunProduksi()).length() + 2){
            kolom[3] = to_string(car.getTahunProduksi()).length() + 2;
        }
        if(kolom[4] < car.getJenisCar().length() + 2){
            kolom[4] = car.getJenisCar().length() + 2;
        }
        if(kolom[5] < to_string(car.getJumlahKursi()).length() + 2){
            kolom[5] = to_string(car.getJumlahKursi()).length() + 2;
        }
        if(kolom[6] < to_string(car.getJumlahPintu()).length() + 2){
            kolom[6] = to_string(car.getJumlahPintu()).length() + 2;
        }
        
    }


    void cek_kolom_motor(Motorcycle motorcycle){
        if(kolom[0] < motorcycle.getPlatNomor().length() + 2){
            kolom[0] = motorcycle.getPlatNomor().length() + 2;
        }
        if(kolom[1] < motorcycle.getMerk().length() + 2){
            kolom[1] = motorcycle.getMerk().length() + 2;
        }
        if(kolom[2] < motorcycle.getWarna().length() + 2){
            kolom[2] = motorcycle.getWarna().length() + 2;
        }
        if(kolom[3] < to_string(motorcycle.getTahunProduksi()).length() + 2){
            kolom[3] = to_string(motorcycle.getTahunProduksi()).length() + 2;
        }
        
        if(kolom[4] < motorcycle.getJenisMotor().length() + 2){
            kolom[4] = motorcycle.getJenisMotor().length() + 2;
        }
        if(kolom[5] < to_string(motorcycle.getKapasitasTangki()).length() + 2){
            kolom[5] = to_string(motorcycle.getKapasitasTangki()).length() + 2;
        }
        
    }

    // Metode untuk menampilkan tabel dengan data yang diberikan.
    void Tampilan_tabel_Car(vector<string> nama_header, vector<Car> data, string nama_tabel) {
        // Inisialisasi variabel i dan j untuk penggunaan dalam loop.
        int i = 0, j = 0;

        // Mengatur lebar kolom berdasarkan panjang header dan data.
        for (i = 0; i < nama_header.size(); i++) {
            if (kolom[i] < (nama_header[i].length() + 2)) {
                kolom[i] = nama_header[i].length() + 2;
            }
            
            
        }

        // Menghitung panjang total kolom.
        int panjang_kolom = 0;
        for (i = 0; i < kolom.size(); i++) {
            panjang_kolom += kolom[i];
        }
        panjang_kolom += 6; // Menambahkan ruang untuk batas tabel.

        // Mencetak batas atas tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;

        // Mencetak nama tabel di tengah.
        cout << "|";
        for (i = 0; i < (panjang_kolom - nama_tabel.length()) / 2; i++) {
            cout << " ";
        }
        cout << nama_tabel;
        for (i = 0; i < (panjang_kolom - nama_tabel.length()) / 2; i++) {
            cout << " ";
        }
        if ((panjang_kolom - nama_tabel.length()) % 2 == 1) {
            cout << " ";
        }
        cout << "|" << endl;

        // Mencetak batas atas nama tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;

        // Mencetak header tabel.
        cout << "| ";
        for (i = 0; i < nama_header.size(); i++) {
            cout << nama_header[i];
            for (j = 0; j < (kolom[i] - 2) - nama_header[i].length(); j++) {
                cout << " ";
            }
            if (i + 1 != nama_header.size()) {
                cout << " | ";
            }
        }
        cout << " |" << endl;

        // Mencetak batas atas header tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;

        // Mencetak data ke dalam tabel.
        for (i = 0; i < data.size(); i++) {
            cout << "| ";
            // Melakukan pencetakan data setiap kolom dengan padding yang sesuai.
            cout << data[i].getPlatNomor();
            for (j = 0; j < (kolom[0] - 2) - data[i].getPlatNomor().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getMerk();
            for (j = 0; j < (kolom[1] - 2) - data[i].getMerk().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getWarna();
            for (j = 0; j < (kolom[2] - 2) - data[i].getWarna().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getTahunProduksi();
            for (j = 0; j < (kolom[3] - 2) - to_string(data[i].getTahunProduksi()).length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getJenisCar();
            for (j = 0; j < (kolom[4] - 2) - data[i].getJenisCar().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getJumlahKursi();
            for (j = 0; j < (kolom[5] - 2) - to_string(data[i].getJumlahKursi()).length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getJumlahPintu();
            for (j = 0; j < (kolom[6] - 2) - to_string(data[i].getJumlahPintu()).length(); j++) {
                cout << " ";
            }
            cout << " |" << endl;
        }

        // Mencetak batas bawah tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;
    }

    void Tampilan_tabel_Motorcycle(vector<string> nama_header, vector<Motorcycle> data, string nama_tabel) {
        // Inisialisasi variabel i dan j untuk penggunaan dalam loop.
        int i = 0, j = 0;

        // Mengatur lebar kolom berdasarkan panjang header dan data.
        for (i = 0; i < nama_header.size(); i++) {
            if (kolom[i] < (nama_header[i].length() + 2)) {
                kolom[i] = nama_header[i].length() + 2;
            }
            
        }

        // Menghitung panjang total kolom.
        int panjang_kolom = 0;
        for (i = 0; i < nama_header.size(); i++) {
            panjang_kolom += kolom[i];
        }
        panjang_kolom += 5; // Menambahkan ruang untuk batas tabel.

        // Mencetak batas atas tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;

        // Mencetak nama tabel di tengah.
        cout << "|";
        for (i = 0; i < (panjang_kolom - nama_tabel.length()) / 2; i++) {
            cout << " ";
        }
        cout << nama_tabel;
        for (i = 0; i < (panjang_kolom - nama_tabel.length()) / 2; i++) {
            cout << " ";
        }
        if ((panjang_kolom - nama_tabel.length()) % 2 == 1) {
            cout << " ";
        }
        cout << "|" << endl;

        // Mencetak batas atas nama tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;

        // Mencetak header tabel.
        cout << "| ";
        for (i = 0; i < nama_header.size(); i++) {
            cout << nama_header[i];
            for (j = 0; j < (kolom[i] - 2) - nama_header[i].length(); j++) {
                cout << " ";
            }
            if (i + 1 != nama_header.size()) {
                cout << " | ";
            }
        }
        cout << " |" << endl;

        // Mencetak batas atas header tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;

        // Mencetak data ke dalam tabel.
        for (i = 0; i < data.size(); i++) {
            cout << "| ";
            // Melakukan pencetakan data setiap kolom dengan padding yang sesuai.
            cout << data[i].getPlatNomor();
            for (j = 0; j < (kolom[0] - 2) - data[i].getPlatNomor().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getMerk();
            for (j = 0; j < (kolom[1] - 2) - data[i].getMerk().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getWarna();
            for (j = 0; j < (kolom[2] - 2) - data[i].getWarna().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getTahunProduksi();
            for (j = 0; j < (kolom[3] - 2) - to_string(data[i].getTahunProduksi()).length(); j++) {
                cout << " ";
            }
            cout << " | ";


            cout << data[i].getJenisMotor();
            for (j = 0; j < (kolom[4] - 2) - data[i].getJenisMotor().length(); j++) {
                cout << " ";
            }
            cout << " | ";

            cout << data[i].getKapasitasTangki();
            for (j = 0; j < (kolom[5] - 2) - to_string(data[i].getKapasitasTangki()).length(); j++) {
                cout << " ";
            }
            
            cout << " |" << endl;
        }

        // Mencetak batas bawah tabel.
        cout << "|";
        for (i = 0; i < panjang_kolom; i++) {
            cout << "=";
        }
        cout << "|" << endl;
    }
    ~Tampilan(){
        
    }

    
};

File: Python/program/Tampilan.py
# Ini adalah import kelas Car dan Motorcycle yang diperlukan.
from Car import Car
from Motorcycle import Motorcycle

# Ini adalah deklarasi dari kelas Tampilan.
class Tampilan:
    def __init__(self, jumlah):
        # Konstruktor kelas Tampilan.
        self.kolom = [0] * jumlah

    def cek_kolom_mobil(self, car):
        # Metode untuk memeriksa dan mengatur lebar kolom berdasarkan data mobil.
        self.kolom[0] = max(self.kolom[0], len(car.get_plat_nomor()) + 2)
        self.kolom[1] = max(self.kolom[1], len(car.get_merk()) + 2)
        self.kolom[2] = max(self.kolom[2], len(car.get_warna()) + 2)
        self.kolom[3] = max(self.kolom[3], len(str(car.get_tahun_produksi())) + 2)
        self.kolom[4] = max(self.kolom[4], len(car.get_jenis_car()) + 2)
        self.kolom[5] = max(self.kolom[5], len(str(car.get_jumlah_kursi())) + 2)
        self.kolom[6] = max(self.kolom[6], len(str(car.get_jumlah_pintu())) + 2)

    def cek_kolom_motor(self, motorcycle):
        # Metode untuk memeriksa dan mengatur lebar kolom berdasarkan data motor.
        self.kolom[0] = max(self.kolom[0], len(motorcycle.get_plat_nomor()) + 2)
        self.kolom[1] = max(self.kolom[1], len(motorcycle.get_merk()) + 2)
        self.kolom[2] = max(self.kolom[2], len(motorcycle.get_warna()) + 2)
        self.kolom[3] = max(self.kolom[3], len(str(motorcycle.get_tahun_produksi())) + 2)
        self.kolom[4] = max(self.kolom[4], len(motorcycle.get_jenis_motor()) + 2)
        self.kolom[5] = max(self.kolom[5], len(str(motorcycle.get_kapasitas_tangki())) + 2)
        

    def Tampilan_tabel_Car(self, nama_header, data, nama_tabel):

        # Metode untuk menampilkan tabel data mobil.
        for i, header in enumerate(nama_header):
            self.kolom[i] = max(self.kolom[i], len(header)+2)
        
        panjang_kolom = sum(self.kolom[:7]) + 6

        # Mencetak batas atas tabel.
        print("|" + "=" * panjang_kolom + "|")
        # Mencetak nama tabel di tengah.
        print("|" + " " * ((panjang_kolom - len(nama_tabel)) // 2) + nama_tabel + " " * ((panjang_kolom - len(nama_tabel)) // 2),end="")
        if (panjang_kolom - len(nama_tabel)) % 2 == 1:
            print(" |")
        else:
            print("|")
        # Mencetak batas atas nama tabel.
        print("|" + "=" * panjang_kolom + "|")
        # Mencetak header tabel.
        print("|", end=" ")
        for i, header in enumerate(nama_header):
            print(f"{header}{' ' * (self.kolom[i] - (len(header) + 2))} |", end=" ")
        print()
        # Mencetak baris pemisah antara header dan data.
        print("|" + "=" * panjang_kolom + "|")
        # Mencetak data.
        
        for mobil in data:
            print("|", end=" ")
            
            print(f"{mobil.get_plat_nomor()}{' ' * (self.kolom[0] - (len(mobil.get_plat_nomor() )+2))} |", end=" ")
            
            print(f"{mobil.get_merk()}{' ' * (self.kolom[1] - (len(mobil.get_merk())+2))} |", end=" ")
            print(f"{mobil.get_warna()}{' ' * (self.kolom[2] - (len(mobil.get_warna())+2))} |", end=" ")
            print(f"{mobil.get_tahun_produksi()}{' ' * (self.kolom[3] - (len(str(mobil.get_tahun_produksi()))+2))} |", end=" ")
            print(f"{mobil.get_jenis_car()}{' ' * (self.kolom[4] - (len(mobil.get_jenis_car() )+2))} |", end=" ")
            print(f"{mobil.get_jumlah_kursi()}{' ' * (self.kolom[5] - (len(str(mobil.get_jumlah_kursi()))+2))} |", end=" ")
            print(f"{mobil.get_jumlah_pintu()}{' ' * (self.kolom[6] - (len(str(mobil.get_jumlah_pintu()))+2))} |")
            
            
            
        # Mencetak batas bawah tabel.
        print("|" + "=" * panjang_kolom + "|")

    def Tampilan_tabel_Motorcycle(self, nama_header, data, nama_tabel):
        # Metode untuk menampilkan tabel data motor.
        for i, header in enumerate(nama_header):
            self.kolom[i] = max(self.kolom[i], len(header)+2)
        
        panjang_kolom = sum(self.kolom[:6]) + 5

        # Mencetak batas atas tabel.
        print("|" + "=" * panjang_kolom + "|")
        # Mencetak nama tabel di tengah.
        print("|" + " " * ((panjang_kolom - len(nama_tabel)) // 2) + nama_tabel + " " * ((panjang_kolom - len(nama_tabel)) // 2),end="")
        if (panjang_kolom - len(nama_tabel)) % 2 == 1:
            print(" |")
        else:
            print("|")
        # Mencetak batas atas nama tabel.
        print("|" + "=" * panjang_kolom + "|")
        # Mencetak header tabel.
        print("|", end=" ")
        for i, header in enumerate(nama_header):
            print(f"{header}{' ' * (self.kolom[i] - (len(header) + 2))} |", end=" ")
        print()
        # Mencetak baris pemisah antara header dan data.
        print("|" + "=" * panjang_kolom + "|")
        # Mencetak data.
        
        for motor in data:
            print("|", end=" ")
            print(f"{motor.get_plat_nomor()}{' ' * (self.kolom[0] - (len(motor.get_plat_nomor())+2))} |", end=" ")
            
            
            print(f"{motor.get_merk()}{' ' * (self.kolom[1] - (len(motor.get_merk())+2))} |", end=" ")
            print(f"{motor.get_warna()}{' ' * (self.kolom[2] - (len(motor.get_warna())+2))} |", end=" ")
            print(f"{motor.get_tahun_produksi()}{' ' * (self.kolom[3] - (len(str(motor.get_tahun_produksi()))+2))} |", end=" ")
            print(f"{motor.get_jenis_motor()}{' ' * (self.kolom[4] - (len(motor.get_jenis_motor())+2))} |", end=" ")
            print(f"{motor.get_kapasitas_tangki()}{' ' * (self.kolom[5] - (len(str(motor.get_kapasitas_tangki()))+2))} |")
            
        # Mencetak batas bawah tabel.
        print("|" + "=" * panjang_kolom + "|")
